File: src/common/web/data/entities/resource/ResourcesList.ts
import { Type } from "class-transformer";

import { Resource } from "./Resource";

/**
 * A list of resource folders.
 */
export type ResourceFolders = ResourcesList[];

/**
 * A list of resource files.
 */
export type ResourceFiles = Resource[];

/**
 * A recursive list of resources.
 *
 * Resources are always given in absolute form.
 *
 * @param resource - The current resource path.
 * @param folders - A list of all folders.
 * @param files - A list of all files.
 */
export class ResourcesList {
    public readonly resource: Resource;

    // @ts-ignore
    @Type(() => ResourcesList)
    public readonly folders: ResourceFolders;
    // @ts-ignore
    @Type(() => Resource)
    public readonly files: ResourceFiles;

    public constructor(resource: Resource, folders: ResourceFolders = [], files: ResourceFiles = []) {
        this.resource = resource;

        this.folders = folders;
        this.files = files;
    }
}

File: src/common/py/data/entities/resource/resources_list.py
import typing
from dataclasses import dataclass, field

from dataclasses_json import dataclass_json

from .resource import Resource

ResourceFolders = typing.List["ResourcesList"]
ResourceFiles = typing.List[Resource]


@dataclass_json
@dataclass(kw_only=True)
class ResourcesList:
    """
    A recursive list of resources.

    Notes:
        Resources are always given in absolute form.

    Attributes:
        resource: The current resource path.
        folders: A list of all folders.
        files: A list of all files.
    """

    resource: Resource

    folders: ResourceFolders = field(default_factory=list)
    files: ResourceFiles = field(default_factory=list)
